plugins/trace_api_plugin/compressed_file.cpp

```diff
@@ -0,0 +1,315 @@
+         file.read(reinterpret_cast<char*>(seek_point_map.data()), seek_point_map.size() * sizeof(seek_point_entry));
+
+         // seek to the neareast seek point
+         auto iter = std::lower_bound(seek_point_map.begin(), seek_point_map.end(), (uint64_t)loc, []( const auto& lhs, const auto& rhs ){
+            return std::get<0>(lhs) < rhs;
+         });
+
+         // special case when there is a seek point that is exact
+         if ( iter != seek_point_map.end() && std::get<0>(*iter) == loc ) {
+            file.seek(std::get<1>(*iter));
+            return;
+         }
+
+         // special case when this is before the first seek point
+         if ( iter == seek_point_map.begin() ) {
+            file.seek(0);
+         } else {
+            // if lower bound wasn't exact iter will be one past the seek point we need
+            const auto& seek_pt = *(iter - 1);
+            file.seek(std::get<1>(seek_pt));
+            remaining -= std::get<0>(seek_pt);
+         }
+      } else {
+         file.seek(0);
+      }
+
+      // read up to the expected offset
+      if (remaining > 0) {
+         auto pre_read_buffer = std::vector<char>(remaining);
+         read(pre_read_buffer.data(), pre_read_buffer.size(), file);
+      }
+   }
+
+   z_stream strm;
+   std::vector<uint8_t> compressed_buffer = std::vector<uint8_t>(compressed_buffer_size);
+   std::vector<uint8_t> read_buffer = std::vector<uint8_t>(read_buffer_size);
+   size_t remaining_read_buffer = 0;
+   bool initialized = false;
+   size_t file_size = 0;
+};
+
+compressed_file::compressed_file( fc::path file_path )
+:file_path(std::move(file_path))
+,file_ptr(nullptr)
+,impl(std::make_unique<compressed_file_impl>())
+{
+   impl->file_size = fc::file_size(file_path);
+}
+
+compressed_file::~compressed_file()
+{}
+
+void compressed_file::seek( long loc ) {
+   impl->seek(loc, *file_ptr);
+
+}
+
+void compressed_file::read( char* d, size_t n ) {
+   impl->read(d, n, *file_ptr);
+}
+
+// these are defaulted now that the opaque impl type is known
+//
+compressed_file::compressed_file( compressed_file&& ) = default;
+compressed_file& compressed_file::operator= ( compressed_file&& ) = default;
+
+
+bool compressed_file::process( const fc::path& input_path, const fc::path& output_path, size_t seek_point_stride  ) {
+   if (!fc::exists(input_path)) {
+      throw std::ios_base::failure(std::string("Attempting to create compressed_file from file that does not exist: ") + input_path.generic_string());
+   }
+
+   const size_t input_size = fc::file_size(input_path);
+   if (input_size == 0) {
+      throw std::ios_base::failure(std::string("Attempting to create compressed_file from file that is empty: ") + input_path.generic_string());
+   }
+
+   // subtract 1 to make sure that the truncated division will only create a seek point if there is at least one byte
+   // in the next stride.  So, a file size of N and a stride <= N results in 0 seek points.  N + 1 will have a seek
```

Comment:
wouldn't this be "and a stride >= N results in 0 seep points."?

Reply:
... yes